talk/morelanguage/ranges.tex

```diff
@@ -0,0 +1,93 @@
+\subsection[range]{Ranges}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Ranges}
+  \begin{block}{The range concept}
+    \begin{itemize}
+    \item A range is a group of items that you can iterator over
+    \item It provides \mintinline{cpp}{begin} and \mintinline{cpp}{end} functions
+    \item Ranges concepts are located in \mintinline{cpp}{std::ranges} namespace
+    \item All STL containers are ranges
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{Variations on ranges}
+    There are actually different type of ranges
+    \begin{itemize}
+    \item \mintinline{cpp}{input_range}, \mintinline{cpp}{output_range}, \mintinline{cpp}{bidirectional_range},  \mintinline{cpp}{random_access_range},  \mintinline{cpp}{contiguous_range}
+    \item mapping the different types of iterators
+    \end{itemize}
+  \end{exampleblock}
+\end{frame}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Views}
+  \begin{block}{The view concept}
+    \begin{itemize}
+    \item A \mintinline{cpp}{view} is something that you apply on a range and performs some operation
+    \item A \mintinline{cpp}{view} does not own data
+    \item It's time to copy, move, assign is constant
+    \item \mintinline{cpp}{view}s can be applied with \mintinline{cpp}{|} and are lazy
+    \item They live in the \mintinline{cpp}{std::views} namespace
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{Example}
+    { \scriptsize
+      \begin{cppcode*}{gobble=4}
+        auto const ints = {0,1,2,3,4,5};
+        auto even = [](int i) { return 0 == i % 2; };
+        auto square = [](int i) { return i * i; };
+        auto results = numbers | std::views::filter(even)
+                               | std::views::transform(square);
+        for (auto a : results) { ... } // 0, 4, 16
+      \end{cppcode*}
+    }
+  \end{exampleblock}
+\end{frame}
+
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Many \texttt{view}s are provided by the STL}
```

Comment:
Again, these are range adapters resulting in the corresponding views. E.g. `std::views::drop` is a range adapter, creating the `std::rangers::drop_view` view.